caas/kubernetes/provider/cloud.go

```diff
@@ -92,61 +91,97 @@ func newCloudCredentialFromKubeConfig(reader io.Reader, cloudParams KubeCloudPar
 }
 
 // UpdateKubeCloudWithStorage updates the passed Cloud with storage details retrieved from the clouds' cluster.
-func UpdateKubeCloudWithStorage(k8sCloud *cloud.Cloud, storageParams KubeCloudStorageParams) (string, error) {
-	fail := func(e error) (string, error) {
-		return "", e
-	}
+func UpdateKubeCloudWithStorage(k8sCloud *cloud.Cloud, storageParams KubeCloudStorageParams) (storageMsg string, err error) {
 
 	// Get the cluster metadata so we can see if there's suitable storage available.
-	clusterMetadata, err := storageParams.GetClusterMetadataFunc(storageParams.MetadataChecker)
+	clusterMetadata, err := storageParams.GetClusterMetadataFunc(storageParams)
+
+	defer func() {
+		if err == nil {
+			var workloadSC, operatorSC string
+
+			// Record the operator storage to use.
+			if clusterMetadata.OperatorStorageClass != nil {
+				operatorSC = clusterMetadata.OperatorStorageClass.Name
+				storageMsg += "."
+			} else {
+				if storageMsg == "" {
+					storageMsg += "\nwith "
+				} else {
+					storageMsg += "\n"
+				}
+				storageMsg += fmt.Sprintf("operator storage provisioned by the workload storage class.")
+			}
+
+			if clusterMetadata.NominatedStorageClass != nil {
+				workloadSC = clusterMetadata.NominatedStorageClass.Name
+			}
+
+			if k8sCloud.Config == nil {
+				k8sCloud.Config = make(map[string]interface{})
+			}
+			if _, ok := k8sCloud.Config[WorkloadStorageKey]; !ok {
+				k8sCloud.Config[WorkloadStorageKey] = workloadSC
+			}
+			if _, ok := k8sCloud.Config[OperatorStorageKey]; !ok {
+				k8sCloud.Config[OperatorStorageKey] = operatorSC
+			}
+		}
+	}()
 
 	if err != nil || clusterMetadata == nil {
 		// err will be nil if user hit Ctrl+C.
 		msg := "cannot get cluster metadata"
 		if err != nil {
 			msg = err.Error()
 		}
-		return fail(ClusterQueryError{Message: msg})
+		return "", ClusterQueryError{Message: msg}
 	}
 
-	if storageParams.HostCloudRegion == "" && clusterMetadata.Regions != nil && clusterMetadata.Regions.Size() > 0 {
-		storageParams.HostCloudRegion = cloud.BuildHostCloudRegion(
-			clusterMetadata.Cloud,
-			clusterMetadata.Regions.SortedValues()[0],
-		)
+	if storageParams.HostCloudRegion == "" && clusterMetadata.Cloud != "" {
+		var region string
+		if clusterMetadata.Regions != nil && clusterMetadata.Regions.Size() > 0 {
+			region = clusterMetadata.Regions.SortedValues()[0]
+		}
+		storageParams.HostCloudRegion = cloud.BuildHostCloudRegion(clusterMetadata.Cloud, region)
 	}
-	if storageParams.HostCloudRegion == "" {
-		return fail(ClusterQueryError{})
+	k8sCloud.HostCloudRegion = storageParams.HostCloudRegion
+
+	cloudType, region, err := cloud.SplitHostCloudRegion(k8sCloud.HostCloudRegion)
+	if err == nil {
+		// try to get region if possible.
```

Comment:
this comment doesn't quite make sense. we are not "getting" the region

Reply:
fixed